### PR TITLE
kafka: Connection error handling

### PR DESCRIPTION
**Pull request by avelanarius**
- Added timeouts to `tcp_client` (and `kafka_connection`).
- Implemented handling of (prematurely) terminated TCP connections.
- Implemented "translation" of network exceptions into error codes (as Kafka already sends errors as error codes and there are a few of them specifically dedicated for network errors).

### CODE REVIEW

File: src/kafka/producer/kafka_producer.cc [hunk 1]
```diff
@@ -47,7 +47,7 @@ namespace kafka {
 kafka_producer::kafka_producer(std::string client_id) : _client_id(std::move(client_id)) {}
 
 seastar::future<> kafka_producer::init(std::string server_address, uint16_t port) {
-    auto connection_future = kafka_connection::connect(server_address, port, _client_id).then([this] (auto connection) {
+    auto connection_future = kafka_connection::connect(server_address, port, _client_id, 500).then([this] (auto connection) {
```
Comment (constantfold) [inline src/kafka/producer/kafka_producer.cc:50]: `500` (ms timeout) hardcoded just for now.

File: src/kafka/connection/kafka_connection.hh [hunk 2]
```diff
@@ -127,6 +127,22 @@ public:
         auto request_future = send_request(request, api_version);
         auto response_future = request_future.then([this, api_version] (int32_t correlation_id) {
             return receive_response<RequestType>(correlation_id, api_version);
+        }).handle_exception([] (auto ep) {
+            try {
+                std::rethrow_exception(ep);
+            } catch (seastar::timed_out_error& e) {
+                typename RequestType::response_type response;
+                response._error_code = 7;
```
Comment (constantfold) [inline src/kafka/connection/kafka_connection.hh:135]: Magic numbers will be gone when we merge Michał's error code support.

Comment (StarostaGit) [reply src/kafka/connection/kafka_connection.hh:135]: good

**Review by StarostaGit — APPROVED** [review]
LGTM, let's merge

**Pull request merged**